### PR TITLE
TIP-1225: ProductModelIndexer as Interface

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

https://akeneo.atlassian.net/browse/TIP-1226  

Remove old interfaces from `Akeneo\Pim\Enrichment\Bundle\Elasticsearch\Indexer\ProductModelIndexer` and replace them by a single one new interface `ProductModelIndexerInterface`.  Public methods modification:

| Old methods | New methods
| --------------------------------- | ---
| index($object, ..) | indexFromProductModelCode(string $productModelCode, ..)
| indexAll($objects, ..) | indexFromProductModelCodes(array $productModelCodes, ..)
| remove($objectId, ..) | removeFromProductModelId(string $productModelId, ..)
| removeAll($objects, ..) | removeFromProductModelIds(array $productModelIds, ..)

Change and adapt in entire application and tests.  
Inject the ProductRepository in order to fetch and normalize product. This is a temporary solution, it will be replaced by a specific model (TIP-1227, TIP-1228).  
Remove the injection of `indexType` in `ProductModelIndexer` as it is not used anymore.  

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 1]
```diff
@@ -20,101 +19,124 @@
...
+                self::INDEX_TYPE,
+                $normalizedProductModels,
+                'id',
+                $options['index_refresh'] ?? Refresh::disable()
+            );
+        }
+    }
+
+    /**
+     * Removes the products from both the product index and the product model index.
```
[the remark this answers is not in the document]
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]: Also, why let the comment here, and not for the other methods?

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 2]
```diff
@@ -20,101 +19,124 @@
...
+     * {@inheritdoc}
+     */
+    public function removeFromProductModelId(string $productModelId, array $options = []): void
+    {
+        $this->removeFromProductModelIds([$productModelId], $options);
     }
 
     /**
-     * Removes the product from both the product model index and the product and product model index.
+     * Removes the products from both the product index and the product and product model index.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]: to fix

File: src/Akeneo/Tool/Component/StorageUtils/Indexer/ProductModelIndexerInterface.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+     * @param string[] $productModelCodes
+     * @param array    $options
+     */
+    public function indexFromProductModelCodes(array $productModelCodes, array $options = []): void;
+
+    /**
+     * @param string $productModelId
+     * @param array  $options
+     */
+    public function removeFromProductModelId(string $productModelId, array $options = []): void;
```
Comment (mmetayer) [inline src/Akeneo/Tool/Component/StorageUtils/Indexer/ProductModelIndexerInterface.php]: nope, that's an int here (I'd rather you cast it here than in the callers)

Comment (mmetayer) [reply src/Akeneo/Tool/Component/StorageUtils/Indexer/ProductModelIndexerInterface.php]: BTW, same goes for ProductIndexer

File: tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductIndexerSpec.php [hunk 4]
```diff
@@ -19,7 +19,7 @@ function let(
         NormalizerInterface $normalizer,
         Client $productAndProductModelIndexClient,
         ProductRepositoryInterface $productRepository
-    ) {
+    )     {
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductIndexerSpec.php]: ??

File: tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php [hunk 5]
```diff
@@ -123,72 +121,66 @@ function it_deletes_product_models_from_elasticsearch_index($productAndProductMo
...
     function it_bulk_deletes_product_models_from_elasticsearch_index($productAndProductModelClient)
     {
-        $productAndProductModelClient->bulkDelete('an_index_type_for_test_purpose', ['product_model_40', 'product_model_33'])
+        $productAndProductModelClient->bulkDelete('pim_catalog_product', ['product_model_40', 'product_model_33'])
             ->shouldBeCalled();
 
-        $this->removeAll([40, 33])->shouldReturn(null);
-    }
-
-    function it_indexes_product_models_and_disables_refresh_of_the_index_by_default(
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php]: why remove this test?

Comment (nmarniesse) [reply tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php]: This test is a duplicate with test `it_bulk_indexes_product_models` on line 78: we already test the index refresh is disabled by default.

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelDescendantsSaver.php [hunk 6]
```diff
@@ -55,15 +51,13 @@ public function __construct(
         ProductModelRepositoryInterface $productModelRepository,
         ProductQueryBuilderFactoryInterface $pqbFactory,
         ProductIndexerInterface $productIndexer,
-        BulkIndexerInterface $bulkProductModelIndexer,
-        IndexerInterface $productModelIndexer,
+        ProductModelIndexerInterface $productModelIndexer,
```
[the remark this answers is not in the document]
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelDescendantsSaver.php]: :+1: Fix in https://github.com/akeneo/pim-community-dev/pull/10672/commits/bf2d64477fd3fcbd02c342790c6be6b375fa747c

File: src/Akeneo/Pim/Enrichment/Bundle/Command/IndexProductModelCommand.php [hunk 7]
```diff
@@ -175,7 +182,7 @@ private function index(OutputInterface $output, array $codes): int
         $i = 0;
         $productModelBulk = [];
         $totalProductModelsIndexed = 0;
-        $progressBar = new ProgressBar($output, $totalProductModelsIndexed);
+        $progressBar = new ProgressBar($output, $productModelsCount);
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Command/IndexProductModelCommand.php]: FYI (I don't care and did this error everytime)
https://english.stackexchange.com/questions/349475/singular-or-plural-noun-preceding-count
;)

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelDescendantsSaver.php [hunk 8]
```diff
@@ -53,17 +43,13 @@ class ProductModelDescendantsSaver implements SaverInterface
 
     public function __construct(
         ProductModelRepositoryInterface $productModelRepository,
-        ProductQueryBuilderFactoryInterface $pqbFactory,
         ProductIndexerInterface $productIndexer,
-        BulkIndexerInterface $bulkProductModelIndexer,
-        IndexerInterface $productModelIndexer,
+        ProductModelIndexerInterface $productModelIndexer,
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelDescendantsSaver.php]: Changelog ?

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 9]
```diff
@@ -20,125 +19,136 @@
...
     ) {
         $this->normalizer = $normalizer;
         $this->productAndProductModelClient = $productAndProductModelClient;
-        $this->indexType = $indexType;
+        $this->productModelRepository = $productModelRepository;
     }
 
     /**
-     * Indexes a product in both the product model index and the product and product model index.
+     * Indexes a product in the product and product model index from it code.
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]: from its*

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/IndexProductModelsSubscriber.php [hunk 10]
```diff
@@ -22,28 +20,15 @@
...
-    private $productModelBulkIndexer;
-
-    /** @var RemoverInterface */
-    private $productModelIndexRemover;
-
     /**
-     * @param IndexerInterface     $productModelIndexer
-     * @param BulkIndexerInterface $productModelBulkIndexer
-     * @param RemoverInterface     $productModelIndexRemover
+     * @param ProductModelIndexerInterface $productModelIndexer
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/IndexProductModelsSubscriber.php]: You can drop construct comments (not mandatory)

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelDescendantsSaver.php [hunk 11]
```diff
@@ -53,17 +43,13 @@ class ProductModelDescendantsSaver implements SaverInterface
 
     public function __construct(
         ProductModelRepositoryInterface $productModelRepository,
-        ProductQueryBuilderFactoryInterface $pqbFactory,
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/Common/Saver/ProductModelDescendantsSaver.php]: need to document that in the CHANGELOG's BC breaks

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 12]
```diff
@@ -20,125 +19,136 @@
...
     ) {
         $this->normalizer = $normalizer;
         $this->productAndProductModelClient = $productAndProductModelClient;
-        $this->indexType = $indexType;
+        $this->productModelRepository = $productModelRepository;
     }
 
     /**
-     * Indexes a product in both the product model index and the product and product model index.
+     * Indexes a product in the product and product model index from its code.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]:
```suggestion
     * Indexes a product model in the product and product model index from its code.
```

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 13]
```diff
@@ -20,125 +19,136 @@
...
-        $this->productAndProductModelClient->index($this->indexType, $normalizedObject['id'], $normalizedObject);
+        $this->indexFromProductModelCodes([$productModelCode], $options);
     }
 
     /**
-     * Indexes a product in both the product model index and the product and product model index.
-     *
-     * If the index_refresh is provided, it uses the refresh strategy defined.
-     * Otherwise the waitFor strategy is by default.
+     * Indexes a list of products in the product and product model index from their codes.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]:
```suggestion
     * Indexes a list of product models in the product and product model index from their codes.
```

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 14]
```diff
@@ -20,125 +19,136 @@
...
+     * {@inheritdoc}
+     */
+    public function removeFromProductModelId(int $productModelId, array $options = []): void
+    {
+        $this->removeFromProductModelIds([$productModelId], $options);
     }
 
     /**
-     * Removes the product from both the product model index and the product and product model index.
+     * Removes the products from the product and product model index.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]:
```suggestion
     * Removes the product models from the product and product model index.
```

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 15]
```diff
@@ -20,125 +19,136 @@
...
+                self::INDEX_TYPE,
+                $normalizedProductModels,
+                'id',
+                $options['index_refresh'] ?? Refresh::disable()
+            );
+        }
+    }
+
+    /**
+     * Removes the product from the product and product model index.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]:
```suggestion
     * Removes the product model from the product and product model index.
```

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php [hunk 16]
```diff
@@ -20,125 +19,136 @@
...
     {
-        $this->productAndProductModelClient->delete(
-            $this->indexType,
-            self::PRODUCT_MODEL_IDENTIFIER_PREFIX . (string) $objectId
-        );
+        if (empty($productModelIds)) {
+            return;
+        }
+
+        if (count($productModelIds) === 1) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelIndexer.php]: not very useful IMO

File: tests/back/Integration/IntegrationTestsBundle/Loader/FixturesLoader.php [hunk 17]
```diff
@@ -389,8 +387,12 @@ protected function indexProducts(): void
      */
     protected function indexProductModels(): void
     {
-        $productModels = $this->container->get('pim_catalog.repository.product_model')->findAll();
-        $this->container->get('pim_catalog.elasticsearch.indexer.product_model')->indexAll($productModels);
+        $this->container->get('pim_catalog.elasticsearch.indexer.product_model')->indexFromProductModelCodes(
```
Comment (mmetayer) [inline tests/back/Integration/IntegrationTestsBundle/Loader/FixturesLoader.php]: :+1:

File: tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php [hunk 18]
```diff
@@ -33,87 +30,88 @@ function it_is_initializable()
 
     function it_is_an_indexer()
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php]:
```suggestion
    function it_is_a_product_model_indexer()
```

File: tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php [hunk 19]
```diff
@@ -33,87 +30,88 @@ function it_is_initializable()
...
+        $productModelRepository->findOneByIdentifier($code)->willReturn($productModel);
         $normalizer->normalize($productModel, ProductAndProductModel\ProductModelNormalizer::INDEXING_FORMAT_PRODUCT_AND_MODEL_INDEX)
-            ->willReturn(['id' => 'baz']);
-        $normalizer->normalize($aWrongProductModel, ProductAndProductModel\ProductModelNormalizer::INDEXING_FORMAT_PRODUCT_AND_MODEL_INDEX)
-            ->willReturn([]);
-
-        $productAndProductModelClient->bulkIndexes(Argument::cetera())->shouldNotBeCalled();
-
-        $this->shouldThrow(\InvalidArgumentException::class)->during('indexAll', [[$productModel, $aWrongProductModel]]);
+            ->willReturn(['id' => $code, 'a key' => 'a value']);
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php]: here 'id' and 'code' are not the same (id is an int whereas code is a string). It does not really matter for the test but fixing that would avoid potential future misunderstanding

File: tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php [hunk 20]
```diff
@@ -33,87 +30,88 @@ function it_is_initializable()
...
-        $this->index($productModel);
+        $code = 'foobar';
+        $productModelRepository->findOneByIdentifier($code)->willReturn(null);
+        $normalizer->normalize(null, ProductAndProductModel\ProductModelNormalizer::INDEXING_FORMAT_PRODUCT_AND_MODEL_INDEX)
+            ->shouldNotBeCalled();
+        $productAndProductModelClient->index(ProductModelIndexer::INDEX_TYPE, $code, ['id' => $code, 'a key' => 'a value'])
+            ->shouldNotBeCalled();
+        $productAndProductModelClient->bulkIndexes(
+            ProductModelIndexer::INDEX_TYPE,
+            [['id' => $code, 'a key' => 'a value']],
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Elasticsearch/Indexer/ProductModelIndexerSpec.php]: same

**Pull request merged**